control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs

```diff
@@ -130,61 +135,139 @@ async fn nexus_reconciler(
     }
 }
 
-/// Find and removes faulted children from the given nexus
-/// If the child is a replica it also disowns and destroys it
+/// Checks if nexus is Degraded and any child is Faulted. If yes, Depending on rebuild policy for
+/// child it performs rebuild operation. We exclude NoSpace Degrade.
 #[tracing::instrument(skip(nexus, context), level = "trace", fields(nexus.uuid = %nexus.uuid(), request.reconcile = true))]
-pub(super) async fn faulted_children_remover(
+pub(super) async fn handle_faulted_child(
     nexus: &mut OperationGuardArc<NexusSpec>,
     context: &PollContext,
 ) -> PollResult {
     let nexus_uuid = nexus.uuid();
     let nexus_state = context.registry().nexus(nexus_uuid).await?;
     let child_count = nexus_state.children.len();
-
-    // Remove faulted children only from a degraded nexus with other healthy children left
     if nexus_state.status == NexusStatus::Degraded && child_count > 1 {
-        let span = tracing::info_span!("faulted_children_remover", nexus.uuid = %nexus_uuid, request.reconcile = true);
-        async {
-            let nexus_spec_clone = nexus.lock().clone();
-            for child in nexus_state.children.iter().filter(|c| c.state.faulted()) {
-                nexus_spec_clone.warn_span(|| {
-                    tracing::warn!(%child.state, %child.state_reason, "Attempting to remove faulted child '{}'", child.uri)
-                });
-                if let Err(error) = nexus
-                    .remove_child_by_uri(
-                        context.registry(),
-                        &nexus_state,
-                        &child.uri,
-                        true,
-                    )
+        let nexus_spec_clone = nexus.lock().clone();
+        for child in nexus_state.children.iter().filter(|c| c.state.faulted()) {
+            if child.state_reason != ChildStateReason::NoSpace {
+                let wait_time = RuleSet::faulted_child_wait(&nexus_state, context.registry());
+                if wait_time.is_zero()
+                    || child.state_reason.clone() == ChildStateReason::RebuildFailed
+                {
+                    info!(%child.uri, "Start full rebuild for child");
+                    faulted_children_remover(nexus, &child.uri, context).await?
+                } else if is_time_elapsed(child.faulted_at, Some(wait_time), &child.uri).await {
+                    info!(
+                        %child.uri, "Start full rebuild for child,  Wait time has elapsed"
+
+                    );
+                    faulted_children_remover(nexus, &child.uri, context).await?;
+                } else if get_child_replica(nexus_spec_clone.clone(), &child.clone(), context)
```

Comment:
I'm sure there is some compact and fancy way in rust to make these two branches look cleaner. Please try. @tiagolobocastro , any suggestion?

Reply:
both else if check different conditions. Hence seemed like an optimal way. You might feel there is lots of branching as we have removed stages. We are checking all possiblities in single reconciler now.

Reply:
Ive made a separate fn for wait now